interfaces/apparmor/backend.go

```diff
@@ -114,74 +113,75 @@ func (b *Backend) Remove(snapName string) error {
 }
 
 var (
-	templatePattern          = regexp.MustCompile("(###[A-Z]+###)")
-	placeholderVar           = []byte("###VAR###")
-	placeholderSnippets      = []byte("###SNIPPETS###")
-	placeholderProfileAttach = []byte("###PROFILEATTACH###")
-	attachPattern            = regexp.MustCompile(`\(attach_disconnected\)`)
-	attachComplain           = []byte("(attach_disconnected,complain)")
+	templatePattern = regexp.MustCompile("(###[A-Z]+###)")
+	attachPattern   = regexp.MustCompile(`\(attach_disconnected\)`)
 )
 
-// combineSnippets combines security snippets collected from all the interfaces
-// affecting a given snap into a content map applicable to EnsureDirState. The
-// backend delegates writing those files to higher layers.
-func (b *Backend) combineSnippets(snapInfo *snap.Info, opts interfaces.ConfinementOptions, snippets map[string][][]byte) (content map[string]*osutil.FileState, err error) {
+const (
+	placeholderVar           = "###VAR###"
+	placeholderSnippets      = "###SNIPPETS###"
+	placeholderProfileAttach = "###PROFILEATTACH###"
+	attachComplain           = "(attach_disconnected,complain)"
+)
+
+func (b *Backend) deriveContent(spec *Specification, snapInfo *snap.Info, opts interfaces.ConfinementOptions) (content map[string]*osutil.FileState, err error) {
 	for _, appInfo := range snapInfo.Apps {
 		if content == nil {
 			content = make(map[string]*osutil.FileState)
 		}
-		addContent(appInfo.SecurityTag(), snapInfo, opts, snippets, content)
+		securityTag := appInfo.SecurityTag()
+		addContent(securityTag, snapInfo, opts, spec.SnippetForTag(securityTag), content)
 	}
 
 	for _, hookInfo := range snapInfo.Hooks {
 		if content == nil {
 			content = make(map[string]*osutil.FileState)
 		}
-		addContent(hookInfo.SecurityTag(), snapInfo, opts, snippets, content)
+		securityTag := hookInfo.SecurityTag()
+		addContent(securityTag, snapInfo, opts, spec.SnippetForTag(securityTag), content)
 	}
 
 	return content, nil
 }
 
-func addContent(securityTag string, snapInfo *snap.Info, opts interfaces.ConfinementOptions, snippets map[string][][]byte, content map[string]*osutil.FileState) {
-	var policy []byte
+func addContent(securityTag string, snapInfo *snap.Info, opts interfaces.ConfinementOptions, snippetForTag string, content map[string]*osutil.FileState) {
+	var policy string
 	if opts.Classic && !opts.JailMode {
 		policy = classicTemplate
 	} else {
 		policy = defaultTemplate
 	}
 	if (opts.DevMode || opts.Classic) && !opts.JailMode {
-		policy = attachPattern.ReplaceAll(policy, attachComplain)
+		policy = attachPattern.ReplaceAllString(policy, attachComplain)
 	}
-	policy = templatePattern.ReplaceAllFunc(policy, func(placeholder []byte) []byte {
+	policy = templatePattern.ReplaceAllStringFunc(policy, func(placeholder string) string {
 		switch {
-		case bytes.Equal(placeholder, placeholderVar):
+		case placeholder == placeholderVar:
```

Comment:
This can be simpler as:
```
switch placeholder {
case "###VAR###":
        ...
case "###SNIPPETS###":
        ...
... etc..
}
```
and then we can kill the global consts. They're used only once, and in this place. So they're obscuring the code without much benefit.